File: scripts/generate-icons.mjs
/**
 * Marka ikonlarını ve sosyal kart görselini üretir.
 * Tek seferlik: `node scripts/generate-icons.mjs` → public/ altına PNG'ler.
 * (sharp yalnızca bu script için geçici devDependency'dir.)
 */
import sharp from 'sharp';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { mkdirSync } from 'node:fs';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const pub = join(root, 'public');
mkdirSync(pub, { recursive: true });

const GRAD = `<linearGradient id="g" x1="0" y1="0" x2="1" y2="1">
  <stop offset="0" stop-color="#4f46e5"/><stop offset="0.55" stop-color="#6366f1"/><stop offset="1" stop-color="#7c3aed"/>
</linearGradient>`;
// 512x512 uzayında ortalanmış beş köşeli yıldız.
const STAR = 'M256 104l44.6 108.4L418 222l-91 74.5L354 412l-98-61.8L158 412l27-115.5-91-74.5 117.4-9.6z';

/** 512x512 viewBox; sharp çıktı boyutuna resize eder. */
function iconSvg({ radius = 0.22, star = 1, bg = true } = {}) {
  const r = Math.round(512 * radius);
  const off = (512 - 512 * star) / 2;
  return `<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="0 0 512 512">
    <defs>${GRAD}</defs>
    ${bg ? `<rect width="512" height="512" rx="${r}" fill="url(#g)"/>` : ''}
    <g transform="translate(${off} ${off}) scale(${star})"><path d="${STAR}" fill="#fff"/></g>
  </svg>`;
}

/** Maskable — dairesel maskeye karşı güvenli alan (yıldız %64, tam kenarlı zemin). */
function maskableSvg() {
  const s = 0.64;
  const off = (512 - 512 * s) / 2;
  return `<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="0 0 512 512">
    <defs>${GRAD}</defs>
    <rect width="512" height="512" fill="url(#g)"/>
    <g transform="translate(${off} ${off}) scale(${s})"><path d="${STAR}" fill="#fff"/></g>
  </svg>`;
}

function ogSvg() {
  return `<svg xmlns="http://www.w3.org/2000/svg" width="1200" height="630" viewBox="0 0 1200 630">
    <defs>
      <linearGradient id="bg" x1="0" y1="0" x2="1" y2="1"><stop offset="0" stop-color="#0b0c10"/><stop offset="1" stop-color="#111827"/></linearGradient>
      ${GRAD}
    </defs>
    <rect width="1200" height="630" fill="url(#bg)"/>
    <circle cx="1080" cy="60" r="380" fill="#6366f1" opacity="0.12"/>
    <rect x="96" y="150" width="128" height="128" rx="30" fill="url(#g)"/>
    <g transform="translate(120 174) scale(0.172)"><path d="${STAR}" fill="#fff"/></g>
    <text x="248" y="242" font-family="system-ui,'Segoe UI',Roboto,sans-serif" font-size="56" font-weight="800" fill="#ffffff">Snaps</text>
    <text x="96" y="386" font-family="system-ui,'Segoe UI',Roboto,sans-serif" font-size="70" font-weight="800" fill="#ffffff">KPSS &amp; YKS AI Sınav Koçu</text>
    <text x="96" y="452" font-family="system-ui,'Segoe UI',Roboto,sans-serif" font-size="33" fill="#94a3b8">Fotoğrafla çöz  ·  Kişisel plan &amp; analiz  ·  Zinciri kırma</text>
  </svg>`;
}

const jobs = [
  ['icon-192.png', iconSvg(), [192, 192]],
  ['icon-512.png', iconSvg(), [512, 512]],
  ['icon-maskable-512.png', maskableSvg(), [512, 512]],
  ['apple-touch-icon.png', iconSvg({ radius: 0, star: 0.82 }), [180, 180]],
  ['favicon-32.png', iconSvg({ radius: 0.28, star: 0.92 }), [32, 32]],
  ['favicon-16.png', iconSvg({ radius: 0.28, star: 0.92 }), [16, 16]],
  ['og-image.png', ogSvg(), [1200, 630]],
];

for (const [name, svg, [w, h]] of jobs) {
  await sharp(Buffer.from(svg), { density: 384 })
    .resize(w, h, { fit: 'cover' })
    .png({ compressionLevel: 9 })
    .toFile(join(pub, name));
  const meta = await sharp(join(pub, name)).metadata();
  console.log('✓', name, `${meta.width}x${meta.height}`);
}

File: scripts/generate-icons.py
#!/usr/bin/env python3
"""
Marka ikonlarını tek kaynaktan üretir: assets/snaps-icon-1024.png (tam taşma, 1024x1024).
Çalıştır:  python3 scripts/generate-icons.py
Çıktı:     public/ altına favicon'lar, apple-touch-icon, PWA ikonları, maskable, og-image.

Bağımlılık: Pillow (yalnızca bu script).
"""
from __future__ import annotations
import os
from PIL import Image, ImageDraw, ImageFilter, ImageFont

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SRC = os.path.join(ROOT, "assets", "snaps-icon-1024.png")
PUB = os.path.join(ROOT, "public")
os.makedirs(PUB, exist_ok=True)

master = Image.open(SRC).convert("RGB")
assert master.size == (1024, 1024), f"kaynak 1024x1024 olmalı, {master.size} bulundu"

LANCZOS = Image.Resampling.LANCZOS


def resized(size: int, crop: float = 0.0) -> Image.Image:
    """crop: her kenardan kırpılacak oran (favicon'da glyph daha büyük görünsün diye)."""
    im = master
    if crop:
        m = round(1024 * crop)
        im = im.crop((m, m, 1024 - m, 1024 - m))
    return im.resize((size, size), LANCZOS)


def save_png(img: Image.Image, name: str) -> None:
    img.save(os.path.join(PUB, name), "PNG", optimize=True)
    print(f"  ✓ {name}  {img.size[0]}x{img.size[1]}")


# --- düz kare ikonlar (tam taşma; platform kendi maskesini uygular) ------------
save_png(resized(512), "icon-512.png")
save_png(resized(192), "icon-192.png")
save_png(resized(180), "apple-touch-icon.png")
# uygulama içi küçük marka işareti (Header / Onboarding) — köşeler kırpılı,
# CSS köşe yuvarlatması temiz otursun diye
save_png(resized(192, crop=0.06), "icon-mark.png")
save_png(resized(32, crop=0.05), "favicon-32.png")
save_png(resized(16, crop=0.05), "favicon-16.png")

# --- favicon.ico (16 / 32 / 48) ----------------------------------------------
resized(256, crop=0.05).save(
    os.path.join(PUB, "favicon.ico"), sizes=[(16, 16), (32, 32), (48, 48)]
)
print("  ✓ favicon.ico  16/32/48")


# --- maskable: içeriği küçült + düz degrade kenarı dışa uzat ----------------
# Önce yuvarlak köşeleri kırp (kenar = düz degrade), sonra S'i %84'e küçültüp
# ortala ki agresif dairesel maske bile kuyruğu/kıvılcımı kesmesin; boşluk
# kenar pikseli uzatılarak dikişsiz doldurulur.
def maskable(size: int, name: str, content: float = 0.84) -> None:
    src = master.crop((60, 60, 964, 964))  # düz degrade kenar
    inner = round(size * content)
    off = (size - inner) // 2
    fg = src.resize((inner, inner), LANCZOS)
    c = Image.new("RGB", (size, size))
    c.paste(fg, (off, off))
    c.paste(fg.crop((0, 0, 1, inner)).resize((off, inner)), (0, off))
    c.paste(fg.crop((inner - 1, 0, inner, inner)).resize((size - off - inner, inner)), (off + inner, off))
    c.paste(c.crop((0, off, size, off + 1)).resize((size, off)), (0, 0))
    c.paste(c.crop((0, off + inner - 1, size, off + inner)).resize((size, size - off - inner)), (0, off + inner))
    save_png(c, name)


maskable(512, "icon-maskable-512.png")
maskable(192, "icon-maskable-192.png")


# --- sosyal kart (og-image 1200x630) ---------------------------------------
def font(size: int, bold: bool = False):
    for p in (
        f"/System/Library/Fonts/Supplemental/Arial{' Bold' if bold else ''}.ttf",
        "/Library/Fonts/Arial.ttf",
        "/System/Library/Fonts/SFNS.ttf",
    ):
        if os.path.exists(p):
            try:
                return ImageFont.truetype(p, size)
            except Exception:
                pass
    return ImageFont.load_default()


def og_image() -> None:
    W, H = 1200, 630
    img = Image.new("RGB", (W, H), "#0b0c10")

    glow = Image.new("RGB", (W, H), "#0b0c10")
    gd = ImageDraw.Draw(glow)
    gd.ellipse([W - 620, -360, W + 300, 420], fill="#4B3BD6")
    glow = glow.filter(ImageFilter.GaussianBlur(160))
    img = Image.blend(img, glow, 0.5)

    sz = 300
    img.paste(master.resize((sz, sz), LANCZOS), (100, (H - sz) // 2))

    d = ImageDraw.Draw(img)
    x = 452
    d.text((x, 214), "Snaps", font=font(104, bold=True), fill="#ffffff")
    d.text((x, 352), "KPSS & YKS AI Sınav Koçu", font=font(40, bold=True), fill="#c7d2fe")
    d.text((x, 410), "Fotoğrafla çöz · Kişisel plan & analiz · Zinciri kırma",
           font=font(26), fill="#93a4bd")
    save_png(img, "og-image.png")


og_image()
print("Bitti.")
